src/hack_forked/utils/collections/flow_map.ml

```diff
@@ -0,0 +1,441 @@
+        match f None with
+        | None -> Empty
+        | Some data -> singleton x data
+      end
+    | Leaf { v; d } ->
+      (* check *)
+      let c = Ord.compare x v in
+      if c = 0 then
+        match f (Some d) with
+        | None -> empty (* It exists, None means deletion *)
+        | Some data ->
+          if d == data then
+            tree
+          else
+            Leaf { v; d = data }
+      else begin
+        match f None with
+        | None -> tree
+        | Some data ->
+          if c < 0 then
+            sorted_two_nodes_smaller x data tree
+          else
+            sorted_two_nodes_larger tree x data
+      end
+    | Node { l; v; d; r; h } as m ->
+      let c = Ord.compare x v in
+      if c = 0 then
+        match f (Some d) with
+        | None -> internal_merge l r
+        | Some data ->
+          if d == data then
+            m
+          else
+            Node { l; v = x; d = data; r; h }
+      else if c < 0 then
+        let ll = update x f l in
+        if l == ll then
+          m
+        else
+          bal ll v d r
+      else
+        let rr = update x f r in
+        if r == rr then
+          m
+        else
+          bal l v d rr
+
+  let rec split x tree =
+    match tree with
+    | Empty -> (Empty, None, Empty)
+    | Leaf { v; d } ->
+      let c = Ord.compare x v in
+      if c = 0 then
+        (empty, Some d, empty)
+      else if c < 0 then
+        (empty, None, tree)
+      else
+        (tree, None, empty)
+    | Node { l; v; d; r; _ } ->
+      let c = Ord.compare x v in
+      if c = 0 then
+        (l, Some d, r)
+      else if c < 0 then
+        let (ll, pres, rl) = split x l in
+        (ll, pres, join rl v d r)
+      else
+        let (lr, pres, rr) = split x r in
+        (join l v d lr, pres, rr)
+
+  let rec merge f s1 s2 =
+    match (s1, s2) with
+    | (Empty, Empty) -> Empty
+    | (Leaf { v; d }, Empty) ->
+      begin
+        match f v (Some d) None with
+        | None -> empty
+        | Some data -> Leaf { v; d = data }
+      end
+    | (Empty, Leaf { v; d }) ->
+      begin
+        match f v None (Some d) with
+        | None -> empty
+        | Some data -> Leaf { v; d = data }
+      end
+    | (Leaf { v = v1; d = d1 }, Leaf _) ->
+      let (l2, d2, r2) = split v1 s2 in
+      concat_or_join (merge f empty l2) v1 (f v1 (Some d1) d2) (merge f empty r2)
+    | (Node { l = l1; v = v1; d = d1; r = r1; h = h1 }, _) when h1 >= height s2 ->
+      let (l2, d2, r2) = split v1 s2 in
+      concat_or_join (merge f l1 l2) v1 (f v1 (Some d1) d2) (merge f r1 r2)
+    | (_, Node { l = l2; v = v2; d = d2; r = r2; _ }) ->
+      let (l1, d1, r1) = split v2 s1 in
+      concat_or_join (merge f l1 l2) v2 (f v2 d1 (Some d2)) (merge f r1 r2)
+    | (Node _, (Empty | Leaf _)) -> assert false
+
+  let rec union f s1 s2 =
+    match (s1, s2) with
+    | (Empty, s)
+    | (s, Empty) ->
+      s
+    | (s, Leaf { v; d })
+    | (Leaf { v; d }, s) ->
+      add v d s
```

Comment:
I believe the behavior in these cases diverges from the existing one. The fact that we're not using `f` at all here is a bad smell.

This repros in the following code 
```
declare var x: {y?: number};
if (x.y === undefined || x.y === null) {
    (x.y: number);
}
```
due to this call to `Key_map.union`: 

https://github.com/facebook/flow/blob/main/src/typing/statement.ml#L7171

Before, the resulting map would bind `x.y` to the entry of the first branch, but after `x.y` gets the entry of the second one.